src/segmentmytif/main.py

```diff
@@ -1,33 +1,88 @@
-                                               feature_type=FeatureType.FLAIR, features_path: Path = None,
-                                               **extractor_kwargs) -> None:
+def read_input_and_labels_and_save_predictions(
+    raster_path: Path,
+    pos_labels_path: Path,
+    neg_labels_path: Path,
+    output_path: Path,
+    feature_type=FeatureType.IDENTITY,
+    features_path: Path = None,
+    mode: Literal["normal", "parallel", "safe"] = "normal",
+    chunks: dict = None,
+    chunk_overlap: int = 16,
+    **extractor_kwargs,
+) -> None:
     logger.info("read_input_and_labels_and_save_predictions called with the following arguments:")
     for k, v in locals().items():
         logger.info(f"{k}: {v}")
 
-    input_data, profile = read_geotiff(input_path)
-
-    features = get_features(input_data, input_path, feature_type, features_path, profile, **extractor_kwargs)
-
-    labels, _ = read_geotiff(labels_path)
-    prediction_map = make_predictions(features, labels)
-
-    save_tiff(prediction_map, output_path, profile)
+    if chunks is None:
+        chunks = {"band": 1, "x": 1000, "y": 1000}
+
+    match mode:
+        case "normal":
+            raster = rioxarray.open_rasterio(raster_path)  # Load in memory
+        case "parallel":
+            raster = rioxarray.open_rasterio(raster_path, chunks=chunks)
+        case "safe":
+            dask.config.set(scheduler="synchronous")
+            raster = rioxarray.open_rasterio(raster_path, chunks=chunks)
+        case _:
+            msg = f"Invalid mode: {mode}"
+            raise ValueError(msg)
+
+    features = get_features(
+        raster, raster_path, feature_type, features_path, chunk_overlap=chunk_overlap, mode=mode, **extractor_kwargs
+    )
+
+    # Load vector labels as geodataframes, and align CRS with input data
+    pos_gdf = gpd.read_file(pos_labels_path).to_crs(raster.rio.crs)
+    neg_gdf = gpd.read_file(neg_labels_path).to_crs(raster.rio.crs)
+
+    # Get label arrays
+    labels = get_label_array(features, pos_gdf, neg_gdf, mode=mode)
+
+    # Make predictions
+    prediction_map = make_predictions(features.data, labels.data)
+
+    # Save predictions
+    prediction_raster = xr.DataArray(
+            prediction_map,
+            dims=raster.dims,
+            coords={"band": range(prediction_map.shape[0]), "y": raster.y, "x": raster.x},
+        )
+    prediction_raster = prediction_raster.rio.write_crs(raster.rio.crs, inplace=True)
+    prediction_raster = prediction_raster.rio.write_transform(raster.rio.transform(), inplace=True)
+    prediction_raster.attrs = raster.attrs
+    prediction_raster.rio.to_raster(output_path)
+
+    # If successful, return the output path
+    if output_path.exists():
+        return output_path
+
+    # If failed, log an error and return None
+    msg = f"Failed to save predictions to {output_path}"
+    logger.error(msg)
+    return None
```

Comment:
Thanks for the thorough error handling. It shows you care, which is awesome :-)
In this specific case though, I think checking for the output file is not redundant. Wouldn't `rio.to_raster` have raised an error in the case anything went wrong? It would be nice to save a few lines of code for readability and also to reduce a little bit of complexity.
```suggestion
    return output_path
```